presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -532,6 +521,16 @@ public PlanOptimizers(
                 costCalculator,
                 new TranslateExpressions(metadata, sqlParser).rules()));
 
+        // TODO: Do not move other PlanNode to SPI until this rule is moved to the end of logical planning (i.e., where AddExchanges lives)
+        // TODO: The connector optimizer should not change plan shape (computations) at this point util #12960 is landed. (e.g., connector may pushdown filters to table scan but cannot add a new node)
+        // TODO: Run RemoveRedundantIdentityProjections and PruneUnreferencedOutputs once (1) we can have ProjectNode in SPI and (2) have moved the connector optimization above HashGenerationOptimizer
+        builder.add(new ApplyConnectorOptimization(planOptimizerManager.getOptimizers()));
```

Comment:
@highker, it looks like there is a bug here. There might be a race between ConnectorManager initializing the optimizers from the connectors and passing them to ApplyConnectorOptimization. ApplyConnectorOptimization has a null set.

Reply:
Good catch! `ConnectorPlanOptimizerManager` is passed into `ConnectorManager` and `PlanOptimizers`. We can enforce a dependency from `ConnectorManager` to `PlanOptimizers` or something similar.